File: data/derived_cot/rq3_prompting/results/translation/qwen/cpp_to_py/code_output/UrlPath.py
import re


class UrlPath:
    def __init__(self):
        self.segments = []
        self._with_end_tag = False

    def add(self, segment):
        self.segments.append(self._fix_path(segment))

    def parse(self, path, charset):
        if path:
            if path[-1] == '/':
                self._with_end_tag = True

            fixed_path = self._fix_path(path)
            if fixed_path:
                segments = fixed_path.split('/')
                for seg in segments:
                    decoded_seg = ''
                    for char in seg:
                        if char != '%':
                            decoded_seg += char
                    self.segments.append(decoded_seg)

    @staticmethod
    def _fix_path(path):
        if not path:
            return ""
        fixed = path
        if fixed.startswith('/'):
            fixed = fixed[1:]
        if fixed.endswith('/'):
            fixed = fixed[:-1]
        return fixed

    def get_segments(self):
        return self.segments

    def with_end_tag(self):
        return self._with_end_tag

File: data/derived_cot/rq3_prompting/results/translation/qwen/py_to_cpp/code_output/UrlPath.cpp
#include <vector>
#include <string>
#include <sstream>
#include <algorithm>

class UrlPath {
private:
    std::vector<std::string> segments;
    bool with_end_tag;

    static std::string fix_path(const std::string& path) {
        if (path.empty()) {
            return "";
        }

        size_t start = 0;
        while (start < path.size() && path[start] == '/') {
            start++;
        }

        size_t end = path.size() - 1;
        while (end > start && path[end] == '/') {
            end--;
        }

        if (start > end) {
            return "";
        }

        return path.substr(start, end - start + 1);
    }

public:
    UrlPath() : segments(), with_end_tag(false) {}

    void add(const std::string& segment) {
        segments.push_back(fix_path(segment));
    }

    void parse(const std::string& path, const std::string& charset) {
        if (path.empty()) {
            return;
        }

        bool had_end_tag = false;
        if (!path.empty() && path.back() == '/') {
            had_end_tag = true;
        }

        std::string fixed_path = fix_path(path);
        segments.clear();

        if (fixed_path.empty()) {
            this->with_end_tag = had_end_tag;
            return;
        }

        std::istringstream iss(fixed_path);
        std::string seg;
        while (std::getline(iss, seg, '/')) {
            std::string decoded_seg = seg;
            // Percent decoding is not directly available in standard C++, so we'll use a placeholder
            // In practice, you would implement percent decoding here
            segments.push_back(decoded_seg);
        }

        this->with_end_tag = had_end_tag;
    }

    // Static method for fixing paths remains unchanged
};